Brokerages/Zerodha/ZerodhaBrokerage.cs

```diff
@@ -0,0 +1,1096 @@
+        /// Gets a list of current subscriptions
+        /// </summary>
+        /// <returns></returns>
+        protected virtual IList<Symbol> GetSubscribed()
+        {
+            IList<Symbol> list = new List<Symbol>();
+            lock (ChannelList)
+            {
+                foreach (var item in ChannelList)
+                {
+                    //TODO: get the security type here
+                    list.Add(Symbol.Create(item.Value.Symbol, SecurityType.Equity, _market));
+                }
+            }
+            return list;
+        }
+
+        public IEnumerable<Slice> GetHistory(IEnumerable<Data.HistoryRequest> requests, DateTimeZone sliceTimeZone)
+        {
+
+            foreach (var request in requests)
+            {
+                if (request.Symbol.ID.SecurityType == SecurityType.Cfd || request.Symbol.ID.SecurityType == SecurityType.Crypto || request.Symbol.ID.SecurityType == SecurityType.Forex || request.Symbol.ID.SecurityType == SecurityType.Commodity)
+                {
+                    throw new ArgumentException("Zerodha does not support this security type: " + request.Symbol.ID.SecurityType);
+                }
+
+                if (request.StartTimeUtc >= request.EndTimeUtc)
+                {
+                    throw new ArgumentException("Invalid date range specified");
+                }
+
+                var start = request.StartTimeUtc.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
+                var end = request.EndTimeUtc.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
+
+                var history = Enumerable.Empty<Slice>();
+
+
+                switch (request.Resolution)
+                {
+                    case Resolution.Tick:
+                        history = GetHistoryTick(request.Symbol, start, end);
+                        break;
+
+                    case Resolution.Second:
+                        history = GetHistorySecond(request.Symbol, start, end);
+                        break;
+
+                    case Resolution.Minute:
+                        history = GetHistoryMinute(request.Symbol, start, end);
+                        break;
+
+                    case Resolution.Hour:
+                        history = GetHistoryHour(request.Symbol, start, end);
+                        break;
+
+                    case Resolution.Daily:
+                        history = GetHistoryDaily(request.Symbol, start, end);
+                        break;
+                }
+
+                foreach (var slice in history)
+                {
+                    yield return slice;
+                }
+            }
+        }
+
+        private IEnumerable<Slice> GetHistoryDaily(Symbol symbol, DateTime start, DateTime end)
+        {
+            Log.Debug("ZerodhaBrokerage.GetHistoryDaily();");
+            string scripSymbol = _symbolMapper.GetBrokerageSymbol(symbol);
+            var candles = _kite.GetHistoricalData(scripSymbol, start, end, "day");
```

Comment:
Seems the Kite REST API calls are rate-limited:
https://kite.trade/forum/discussion/257/api-rate-limiting

See the usages of the `RateGate` class in other brokerages for how to handle rate limits.

Reply:
Will implement

Reply:
Rate limiting is Per API basis only for market instruments and quote API's which we barely use and it is made sure that we are not hitting the API more than once/sec.